sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -353,12 +353,20 @@ Command *Scheduler::GraphBuilder::insertMemoryMove(MemObjRecord *Record,
     Record->MHostAccess = MapMode;
   } else {
 
-    // Full copy of buffer is needed to avoid loss of data that may be caused
-    // by copying specific range from host to device and backwards.
-    NewCmd =
-        new MemCpyCommand(*AllocaCmdSrc->getRequirement(), AllocaCmdSrc,
-                          *AllocaCmdDst->getRequirement(), AllocaCmdDst,
-                          AllocaCmdSrc->getQueue(), AllocaCmdDst->getQueue());
+    if (Req->MAccessMode == access::mode::discard_write) {
+      EmptyCommand *Cmd = new EmptyCommand(AllocaCmdDst->getQueue());
+      Cmd->addRequirement(/* DepCmd = */ nullptr, /* AllocaCmd = */ nullptr,
+                          Req);
+      Cmd->MBlockReason = Command::BlockReason::HostAccessor;
+      NewCmd = Cmd;
+    } else {
+      // Full copy of buffer is needed to avoid loss of data that may be caused
+      // by copying specific range from host to device and backwards.
+      NewCmd =
+          new MemCpyCommand(*AllocaCmdSrc->getRequirement(), AllocaCmdSrc,
+                            *AllocaCmdDst->getRequirement(), AllocaCmdDst,
+                            AllocaCmdSrc->getQueue(), AllocaCmdDst->getQueue());
+    }
   }
```

Comment:
@sergey-semenov  , is this more in line with what you were thinking? I tried putting the check before `insertMemoryMove`, but that's not practical as that can interfere with Map/Unmap.

Reply:
Yeah, that looks similar to what I had in mind. Is the empty command here needed? Is it possible to just change the current context and return `nullptr` for this case?

Reply:
No - I just reattempted just to make sure. A nullptr leads to a segfault when there is no host unified memory. I spent some time on this earlier in the week and IIRC it wasn't a simple fix in one place. The simplest safest solution is to use an empty command.

Reply:
The old "do nothing" MemCpyCommand never "does nothing", it always enqueues a mem operation.   And the reason for that is that here we have two potential requirements to use: `Req` and `AllocaCmdDst->getRequirement()` and `Req` has the discard_write, but the second does not.   So the old discard_write check/"do nothing" is never taken because it is never passed a requirement with that access mode.    So that's the difference between this PR and what we have now. (Does that make sense?)

If we use the `Req` argument, then the behavior vis-a-vis discard_write is correct.  However we get a different failure when image mem objects are used. And that has to do with some strange (potential bug) initialization occurring over in the imageaccessors.

Reply:
And, yes, I'll go hunt down that segfault again and report back.

Reply:
Yes, I should have been clearer: I meant to compare the two approaches to fixing this bug.